### PR TITLE
Chhs data

### PR DESCRIPTION
**Pull request by wh1210**

### CODE REVIEW

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.mcf [hunk 1]
```diff
@@ -0,0 +1,133 @@
+Node: dcid:Count_Hosptital_GeneralAcuteCare
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.mcf:1]: s/Hosptital/Hospital/g (occurs a few times in this PR)

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.mcf [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+medicalSpecialty: dcs:GeneralAcuteCare
+bedType: dcs:RenalTransplantBed
+
+Node: dcid:Count_HospitalBed_GeneralAcuteCare_UnspecifiedGeneralAcuteCareBed
+typeOf: dcs:StatisticalVariable
+populationType: dcs:HospitalBed
+statType: dcs:measuredValue
+measuredProperty: dcs:count
+medicalSpecialty: dcs:GeneralAcuteCare
+bedType: dcs:UnspecifiedGeneralAcuteCareBed
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.mcf:133]: Prefer \n as last character of file

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.tmcf [hunk 3]
```diff
@@ -0,0 +1,119 @@
+
+Node: E:CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts->E0
+typeOf: dcs:StatVarObservation
+variableMeasured: dcs:Count_Hosptital_GeneralAcuteCare
+observationDate: "2020-04-13"
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/CA_County_General_Acute_Care_Hospitals_Bed_Types_And_Counts.tmcf:5]: Date should be not hardcoded in template

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,3 @@
+numpy==1.17.4
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/requirements.txt:1]: Can put in the github directory level requirements.txt

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/config.txt [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    # For dataset 1
+    mcf_url: "gs://resolved_mcf/external_tables/ca_chhs_1/nodes.mcf-00000-of-00001"
+    # For dataset 2
+    mcf_url: "gs://resolved_mcf/external_tables/ca_chhs_2_write/nodes.mcf-00000-of-00001"
+
+    # First date of imported data
+    start_date: "2020-06-01"
+    # Last date of imported data, leave it blank for now because there are only one date available, we can update regularly later to get a time sequence data.
+
+    # Source data release schedule. Leave it blank for now as there is no description on website.
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/config.txt:16]: Can remove this comment.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/new_schema.mcf [hunk 6]
```diff
@@ -0,0 +1,170 @@
+# -----------------New Schema------------------------
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/new_schema.mcf:1]: rm this line

Comment (tjann) [reply scripts/us_ca_chhs/healthcare_facility_bed_counts/new_schema.mcf:1]: And put the rest of this file in the corresponding file in the top level schema/ directory

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+import os
+from preprocess_csv_helper import preserve_leading_zero
+from preprocess_csv_helper import generate_dcid_for_county
+
+
+# process dataset 1
+df1 = pd.read_excel('https://data.chhs.ca.gov/dataset/09b8ad0e-aca6-4147-b78d-bdaad872f30b/resource/0997fa8e-ef7c-43f2-8b9a-94672935fa60/download/healthcare_facility_beds.xlsx',
+                    converters={'FACID': lambda x: str(x)})
+
+new_columns = ['GeoId', 'Date',
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:26]: We generally prefer having all the constants defined at the top, before doing any actual calls like pd.read_excel.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+               'Count_HospitalBed_ChemicalDependencyRecoveryBed',
+               'Count_HospitalBed_PediatricIntensiveCareUnitBed',
+               'Count_HospitalBed_LaborAndDeliveryBed',
+               'Count_HospitalBed_IntermediateCareDevelopmentallyDisabledBed',
+               'Count_HospitalBed_PsychiatricHealthBed',
+               'Count_HospitalBed_PediatricDayAndRespiteCareBed',
+               'Count_HospitalBed_DialysisStationsBed',
+               'Count_HospitalBed_CorrectionalTreatmentCenterBed']
+
+new_df = pd.DataFrame(columns=new_columns)
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:56]: This should also be moved to after all the constant declarations.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    county_to_geoID = {}
+    sorted_counties = np.sort(counties)
+    # see browser.datacommons.org/kg?dcid=geoId/06 for convention for geoId to county.
+    counties_id = [i for i in range(1, len(counties) * 2 + 1, 2)]
+    for i in range(len(sorted_counties)):
+        county_to_geoID[sorted_counties[i]] = perserve_length(counties_id[i])
+    return county_to_geoID
+
+
+# A helper function to generate geoId for given county name.
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper.py:29]: Seems to me this function returns the geoId of a county given its fips code. Maybe make the name reflect that? Also I thought all counties have exactly 3 digits, so you could probABLY USE zfill.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+def get_camel_formatting_name_list(name_list):
+    formatting_list = []
+    for name in name_list:
+        formatting_list.append(get_camel_formatting_name(name))
+    return formatting_list
+
+
+def get_camel_formatting_name(name):
+    after_spliter = name.split(' ')
+    after_cap = [spliter.capitalize() for spliter in after_spliter]
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper.py:63]: Might be able to use https://www.w3schools.com/python/ref_string_title.asp and just replace(' ', '') ?

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper_test.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+import preprocess_csv_helper
+
+class PreprocessCSVHelperTest(unittest.TestCase):
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv_helper_test.py:18]: Can you also add tests for the core (non helper function) preprocess_csv.py? May need to modularize the code a bit for easier testing (1 - get data into a data frame; 2 - act on the data frame). I suggest having a testdata subdir where you put a few edge case rows of the data, and read in that dataframe to test the second part.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+    'CHRONIC DIALYSIS CLINIC': 'ChronicDialysisFacility',
+    'CHEMICAL DEPENDENCY RECOVERY HOSPITAL': 'ChemicalDependencyRecoveryFacility',
+    'CORRECTIONAL TREATMENT CENTER': 'CorrectionalTreatmentCenterFacility'
+}
+
+bed_type_dict = {
+    'SKILLED NURSING': 'SkilledNursingBed',
+    'SPECIAL TREATMENT PROGRAM': 'SpecialTreatmentProgramBed',
+    'INTERMEDIATE CARE/DD HABILITATIVE': 'IntermediateCareHabilitativeBed',
+    'INTERMEDIATE CARE/DD NURSING': 'IntermediateCareNursingBed',
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:77]: Can you remind me why we don't parse DD into DevelopmentallyDisabled here? (as opposed to IntermediateCareDevelopmentallyDisabledBed, where we do)

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+    'ACUTE PSYCHIATRIC CARE': 'AcutePsychiatricCareBed',
+    'CHEMICAL DEPENDENCY RECOVERY': 'ChemicalDependencyRecoveryBed',
+    'PEDIATRIC INTENSIVE CARE UNIT': 'PediatricIntensiveCareUnitBed',
+    'LABOR AND DELIVERY': 'LaborAndDeliveryBed',
+    'INTERMEDIATE CARE/DD': 'IntermediateCareDevelopmentallyDisabledBed',
+    'PSYCHIATRIC HEALTH': 'PsychiatricHealthBed',
+    'PEDI. DAY & RESPITE CARE': 'PediatricDayAndRespiteCareBed',
+    'DIALYSIS STATIONS': 'DialysisStationsBed',
+    'CORRECTIONAL TREATMENT CENTER': 'CorrectionalTreatmentCenterBed',
+}
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:100]: I wonder if handling a few special cases, and then just doing a Title Case + strip whitespace would be better than this long map. I will leave this comment for other reviewers to see what they think.

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+            'Count_HospitalBed_AcuteRespiratoryCareBed': None,
+            'Count_HospitalBed_AcutePsychiatricCareBed': None,
+            'Count_HospitalBed_ChemicalDependencyRecoveryBed': None,
+            'Count_HospitalBed_PediatricIntensiveCareUnitBed': None,
+            'Count_HospitalBed_LaborAndDeliveryBed': None,
+            'Count_HospitalBed_IntermediateCareDevelopmentallyDisabledBed': None,
+            'Count_HospitalBed_PsychiatricHealthBed': None,
+            'Count_HospitalBed_PediatricDayAndRespiteCareBed': None,
+            'Count_HospitalBed_DialysisStationsBed': None,
+            'Count_HospitalBed_CorrectionalTreatmentCenterBed': None
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:150]: Can you see if there is a way to add a row to a pd DataFrame without explicitly defining all values (default to None)?

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+
+# observation node, observationAbout facility
+TEMPLATE_MCF_TEMPLATE_1 = """
+Node: E:CA_Licensed_Healthcare_Facility_Types_And_Counts->E{index}
+typeOf: StatVarObservation
+variableMeasured: {stat_var}
+observationDate: C:CA_Licensed_Healthcare_Facility_Types_And_Counts->Date
+observationAbout: E:CA_Licensed_Healthcare_Facility_Types_And_Counts->E0
+value: C:CA_Licensed_Healthcare_Facility_Types_And_Counts->{stat_var}
+"""
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:176]: Also move these declarations towards the top (this will help us when we start to modularize the code)

File: scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+
+with open('CA_Licensed_Healthcare_Facility_Types_And_Counts.tmcf', 'w', newline='') as f_out:
+    f_out.write(TEMPLATE_MCF_FAC_1)
+
+    stat_vars = new_columns[5:]
+    for i in range(len(stat_vars)):
+        f_out.write(TEMPLATE_MCF_TEMPLATE_1.format_map({'index': i + 1, 'stat_var': stat_vars[i]}))
+
+# Proprocess for dataset2 and output tmcf file.
```
Comment (tjann) [inline scripts/us_ca_chhs/healthcare_facility_bed_counts/preprocess_csv.py:185]: Is there a benefit for stuffing both datasets in one script? They seem more like 2 separate imports.

**Comment by beets** [comment]
Hi, any updates here?

**Comment by wh1210** [comment]
> Hi, any updates here?

I will take a look at this soon!
